File: src/exoticatechnologies/ui/impl/shop/upgrades/methods/CreditsMethod.java
package exoticatechnologies.ui.impl.shop.upgrades.methods;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.campaign.econ.MarketAPI;
import com.fs.starfarer.api.fleet.FleetMemberAPI;
import com.fs.starfarer.api.impl.campaign.ids.Commodities;
import com.fs.starfarer.api.util.Misc;
import exoticatechnologies.hullmods.ExoticaTechHM;
import exoticatechnologies.modifications.ShipModLoader;
import exoticatechnologies.modifications.upgrades.Upgrade;
import exoticatechnologies.modifications.ShipModifications;
import exoticatechnologies.util.StringUtils;
import exoticatechnologies.util.Utilities;

import java.util.HashMap;
import java.util.Map;


public class CreditsMethod implements UpgradeMethod {
    @Override
    public String getOptionText(FleetMemberAPI fm, ShipModifications es, Upgrade upgrade, MarketAPI market) {
        int level = es.getUpgrade(upgrade);
        float resourceCreditCost = getCreditCostForResources(upgrade.getResourceCosts(fm, level));
        int convenienceFee = getConvenienceCreditCost(resourceCreditCost, level, upgrade.getMaxLevel(fm.getHullSpec().getHullSize()), market);
        int creditCost = getFinalCreditCost(fm, upgrade, level, market);

        String creditCostFormatted = Misc.getFormat().format(creditCost);
        String convenienceFeeFormatted = Misc.getFormat().format(convenienceFee);
        return StringUtils.getTranslation("UpgradeMethods", "CreditsOption")
                .format("credits", creditCostFormatted)
                .format("extraTax", convenienceFeeFormatted)
                .toString();
    }

    @Override
    public String getOptionTooltip(FleetMemberAPI fm, ShipModifications es, Upgrade upgrade, MarketAPI market) {
        return StringUtils.getString("UpgradeMethods", "CreditsUpgradeTooltip");
    }

    @Override
    public boolean canUse(FleetMemberAPI fm, ShipModifications es, Upgrade upgrade, MarketAPI market) {
        if (upgrade.getResourceRatios().isEmpty()) return false;

        int level = es.getUpgrade(upgrade);
        int creditCost = getFinalCreditCost(fm, upgrade, level, market);

        return Global.getSector().getPlayerFleet().getCargo().getCredits().get() >= creditCost;
    }

    @Override
    public boolean canShow(FleetMemberAPI fm, ShipModifications es, Upgrade upgrade, MarketAPI market) {
        return true;
    }

    @Override
    public String apply(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market) {
        int level = mods.getUpgrade(upgrade);
        int creditCost = getFinalCreditCost(fm, upgrade, level, market);
        Global.getSector().getPlayerFleet().getCargo().getCredits().subtract(creditCost);

        mods.putUpgrade(upgrade);
        ShipModLoader.set(fm, mods);
        ExoticaTechHM.addToFleetMember(fm);

        Global.getSoundPlayer().playUISound("ui_char_increase_skill_new", 1f, 1f);
        return StringUtils.getTranslation("UpgradesDialog", "UpgradePerformedSuccessfully")
                .format("name", upgrade.getName())
                .format("level", mods.getUpgrade(upgrade))
                .toString();
    }

    /**
     * Sums up the floats in the map.
     *
     * @param resourceCosts
     * @return The sum.
     */
    private static int getCreditCostForResources(Map<String, Integer> resourceCosts) {
        float creditCost = 0;

        for (Map.Entry<String, Integer> resourceCost : resourceCosts.entrySet()) {
            creditCost += Utilities.getItemPrice(resourceCost.getKey()) * resourceCost.getValue();
        }
        return (int) creditCost;
    }

    /**
     * A formula that uses the market's relations with the player to determine a "convenience cost" for an upgrade.
     *
     * @param market      the market
     * @param upgradeCost the cost of the upgrade
     * @param level       the level of the upgrade
     * @param max         the max level
     * @return
     */
    private static int getConvenienceCreditCost(float upgradeCost, int level, int max, MarketAPI market) {
        float rel = market.getFaction().getRelToPlayer().getRel();
        float exp = (float) (1 + 4.5 * level / max);
        float base = 2f - 0.5f * rel;
        float additive = (float) (upgradeCost * Math.pow(base, exp));

        return (int) additive;
    }

    /**
     * Calculates the cost of the upgrade based on the resources it uses and an additional convenience cost.
     *
     * @param fm      the ship to upgrade
     * @param upgrade the upgrade
     * @param level   the level of the upgrade
     * @param market  the market
     * @return the cost
     */
    public static int getFinalCreditCost(FleetMemberAPI fm, Upgrade upgrade, int level, MarketAPI market) {
        float creditCost = getCreditCostForResources(upgrade.getResourceCosts(fm, level));
        int convenienceFee = getConvenienceCreditCost(creditCost, level, upgrade.getMaxLevel(fm.getHullSpec().getHullSize()), market);
        return (int) (creditCost + convenienceFee);
    }

    @Override
    public Map<String, Float> getResourceCostMap(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market, boolean hovered) {
        Map<String, Float> resourceCosts = new HashMap<>();

        float cost = 0f;
        if (hovered) {
            cost = (float) CreditsMethod.getFinalCreditCost(fm, upgrade, mods.getUpgrade(upgrade), market);
        }
        resourceCosts.put(Commodities.CREDITS, cost);

        return resourceCosts;
    }

    @Override
    public boolean usesBandwidth() {
        return true;
    }

    @Override
    public boolean usesLevel() {
        return true;
    }
}

File: src/exoticatechnologies/ui/impl/shop/upgrades/methods/CreditsMethod.kt
package exoticatechnologies.ui.impl.shop.upgrades.methods

import com.fs.starfarer.api.Global
import com.fs.starfarer.api.campaign.econ.MarketAPI
import com.fs.starfarer.api.combat.ShipVariantAPI
import com.fs.starfarer.api.fleet.FleetMemberAPI
import com.fs.starfarer.api.impl.campaign.ids.Commodities
import exoticatechnologies.hullmods.ExoticaTechHM
import exoticatechnologies.modifications.ShipModLoader
import exoticatechnologies.modifications.ShipModifications
import exoticatechnologies.modifications.upgrades.Upgrade
import exoticatechnologies.util.StringUtils
import lombok.Getter

class CreditsMethod : DefaultUpgradeMethod() {
    @Getter
    override var key = "credits"
    override fun getOptionText(member: FleetMemberAPI, mods: ShipModifications, upgrade: Upgrade, market: MarketAPI?): String {
        return StringUtils.getTranslation("UpgradeMethods", "CreditsOption")
            .toString()
    }

    override fun getOptionTooltip(
        member: FleetMemberAPI,
        mods: ShipModifications,
        upgrade: Upgrade,
        market: MarketAPI?
    ): String {
        return StringUtils.getString("UpgradeMethods", "CreditsUpgradeTooltip")
    }

    override fun canUse(member: FleetMemberAPI, mods: ShipModifications, upgrade: Upgrade, market: MarketAPI?): Boolean {
        market ?: return false

        if (upgrade.resourceRatios.isEmpty()) return false
        val level = mods.getUpgrade(upgrade)
        val creditCost = getFinalCreditCost(member, upgrade, level, market)
        return (Global.getSector().playerFleet.cargo.credits.get() >= creditCost
                && super.canUse(member, mods, upgrade, market))
    }

    override fun apply(member: FleetMemberAPI, variant: ShipVariantAPI, mods: ShipModifications, upgrade: Upgrade, market: MarketAPI?): String {
        val level = mods.getUpgrade(upgrade)
        val creditCost = getFinalCreditCost(member, upgrade, level, market!!)
        Global.getSector().playerFleet.cargo.credits.subtract(creditCost.toFloat())

        mods.putUpgrade(upgrade)
        ShipModLoader.set(member, variant, mods)
        ExoticaTechHM.addToFleetMember(member, variant)

        return StringUtils.getTranslation("UpgradesDialog", "UpgradePerformedSuccessfully")
            .format("name", upgrade.name)
            .format("level", mods.getUpgrade(upgrade))
            .toString()
    }

    override fun getResourceCostMap(
        member: FleetMemberAPI,
        mods: ShipModifications,
        upgrade: Upgrade,
        market: MarketAPI?,
        hovered: Boolean
    ): Map<String, Float> {
        val resourceCosts: MutableMap<String, Float> = HashMap()
        if (hovered) {
            resourceCosts[Commodities.CREDITS] = getFinalCreditCost(member, upgrade, mods.getUpgrade(upgrade), market!!).toFloat()
        }
        return resourceCosts
    }

    companion object {
        /**
         * A formula that uses the market's relations with the player to determine a "convenience cost" for an upgrade.
         *
         * @param market      the market
         * @param upgradeCost the cost of the upgrade
         * @param level       the level of the upgrade
         * @param max         the max level
         * @return
         */
        private fun getConvenienceCreditCost(upgradeCost: Float, level: Int, max: Int, market: MarketAPI): Int {
            val rel = market.faction.relToPlayer.rel
            val exp = (1 + 4.5 * level / max).toFloat()
            val base = 2f - 0.5f * rel
            val additive = (upgradeCost * Math.pow(base.toDouble(), exp.toDouble())).toFloat()
            return additive.toInt()
        }

        /**
         * Calculates the cost of the upgrade based on the resources it uses and an additional convenience cost.
         *
         * @param fm      the ship to upgrade
         * @param upgrade the upgrade
         * @param level   the level of the upgrade
         * @param market  the market
         * @return the cost
         */
        fun getFinalCreditCost(fm: FleetMemberAPI, upgrade: Upgrade, level: Int, market: MarketAPI): Int {
            val creditCost = upgrade.getCreditCostForResources(upgrade.getResourceCosts(fm, level)).toFloat()
            val convenienceFee =
                getConvenienceCreditCost(creditCost, level, upgrade.maxLevel, market)
            return (creditCost + convenienceFee).toInt()
        }
    }
}